### ISSUE TITLE
Add easier get_child() variations

### ISSUE THREAD

**Issue by JMS55** [gtk-rs/gtk4-rs#239]
A lot of the time I do something like this:
```rust
widget.get_child().unwrap().downcast::<SomeWidget>().unwrap();
```
This gets to be really tedious and ugly. I propose gtk-rs should add a function equivalent to the above, something like:
```rust
widget.get_child_as::<SomeWidget>();
```

**Comment by sdroege** [comment]
Sounds useful, yes

**Comment by bilelmoussaoui** [comment]
What makes it harder is that get_child is not a `WidgetExt` method and we will have to add it to a bunch of various types manually.  Do you want to send a PR for that? I think having a trait with the provided method implemented for all those types would be an easy way to do it.

**Comment by sophie-h** [comment]
Is there something special about `get_child`? Otherwise, I would expect all functions that return a Widget to have a `_as` variant and I'm not sure if it's worth the effort to hard code that into gir.

**Comment by bilelmoussaoui** [comment]
> Is there something special about `get_child`? 

Nothing special about it, there are definitely other functions that could benefit from a generic alternative.  Do you have any others in mind? 

> Otherwise, I would expect all functions that return a Widget to have a `_as` variant and I'm not sure if it's worth the effort to hard code that into gir.

Yeah no, there's no way to detect/generate that automatically from gir

**Comment by sophie-h** [comment]
`Stack::get_visible_child` for example and I probably a bunch of functions in libadwaita like `hdy_action_row_get_activatable_widget ()`. I am a bit scared that there will be too much manual stuff at some point and I'm not sure about cost/benefit with this issue.

**Comment by ricvelozo** [comment]
Maybe is possible to specialize, something like:

```rust
trait OptionCast { /* ... */ } // Similar to Cast trait, but returns the value or panics

impl OptionCast for Option<glib::Object> {
    fn downcast<T: ObjectType>(self) -> T {
        panic!("failed to cast to `{}`", T::static_type())
    }
}

// Now
list_item.get_child().unwrap().downcast::<gtk::Label>().expect("failed to cast to `GtkLabel`");

// After
list_item.get_child().downcast::<gtk::Label>();
```

**Comment by sophie-h** [comment]
@bilelmoussaoui I don't think this needs to be on 0.1.0

**Comment by JMS55** [comment]
Yeah, my idea was to also do it for get_item() and various other functions, but that's a lot of work manually. I'm thinking we would either want to have gir generate a `get_x_as()` function for any function that looks like `get_x() -> Object`, but not sure how feasible that is. The much more realistic option, but perhaps slightly worse to use, is what @ricvelozo posted - Add a new function on the Cast trait or somewhere, that at least lets you get rid of two unwraps.

**Comment by bilelmoussaoui** [comment]
> Maybe is possible to specialize, something like:
> 
> ```rust
> trait OptionCast { /* ... */ } // Similar to Cast trait, but returns the value or panics
> 
> impl OptionCast for Option<glib::Object> {
>     fn downcast<T: ObjectType>(self) -> T {
>         panic!("failed to cast to `{}`", T::static_type())
>     }
> }
> 
> // Now
> list_item.get_child().unwrap().downcast::<gtk::Label>().expect("failed to cast to `GtkLabel`");
> 
> // After
> list_item.get_child().downcast::<gtk::Label>();
> ```

@sdroege What do you think of something like this? without the panic part. I suppose it's similar to the ListModel::item function case where you lose the semantics of the `Option` in case you're trying to cast to a wrong type.

**Comment by sdroege** [comment]
You could implement `Downcast` etc on `Option<T>` and then you'd end up with `fn downcast<U>(Option<T>) -> Result<Option<U>, Option<T>>`. That would not hide the errors at least while still allowing to directly work on `Option`s.

**Comment by bilelmoussaoui** [comment]
I am going to move this to gtk-rs-core as such trait implementation has to be done on glib-rs side
